asyncscope.md

```diff
@@ -2317,7 +2317,225 @@ spec here
+- [1.7]{.pnum} `@_joined_@`: when the count of associated objects drops to
+    zero while `s` is in `@_open-and-joining_@` or `@_closed-and-joining_@`
+    state, `s` moves to the `@_joined_@` state.
+
+##### Constructor and Destructor [exec.scounting.ctor]
+
+`simple_counting_scope() noexcept;`
+
+[1]{.pnum} _Postcondtions:_ `@_count_@` is `0` and `@_state_@` is `@_unused_@`
+
+`~simple_counting_scope();`
+
+[2]{.pnum} _Effects:_ If `@_state_@` is not one of `@_joined_@`, `@_unused_@`,
+    or `@_unused-and-close_@`, invokes `terminate` ([except.terminate]{.sref}).
+    Otherwise, has no effects.
+
+##### Members [exec.scounting.mem]
+
+`token get_token() noexcept;`
+
+[1]{.pnum} _Returns:_ An object `t` of type token such that
+    `t.@_scope_@ == this` is `true`.
+
+`void close() noexcept;`
+
+[2]{.pnum} _Efects:_ If `@_state_@` is
+
+- [2.1]{.pnum} `@_unused_@` changes `@_state_@` to `@_unused-and_close_@`;
+- [2.2]{.pnum} `@_open_@` changes `@_state_@` to `@_closed_@`;
+- [2.3]{.pnum} `@_open-and-joining_@` changes `@_state_@` to `@_closed-and-joining_@`;
+
+[3]{.pnum} Any call to `t.try_associate()` for a `token` object `t`
+    referring to a `simple_counting_scope` object `s` which are
+    sequenced after a call to `s.close()` return a disengaged
+    association ([exec.scounting.assoc]{.sref}, p1).
+
+`auto join() noexcept;`
+
+[4]{.pnum} _Returns:_ `@_make_sender_@(@_join-t_@, this)`
+
+[5]{.pnum} The exposition-only class template `@_impls-for_@`
+    ([exec.snd.general]{.sref}) is specialized for `@_join-t_@`
+    as follows:
+
+```
+    template <>
+    struct @_impls-for_@<@_join-t_@>: @_default-impls_@> {
+        template <class Receiver>
+        struct @_state_@ {  // @_exposition-only_@
+            simple_counting_scope* @_scope_@; // @_exposition-only_@
+            remove_cvref_t<Receiver>& @_receiver_@; // @_exposition-only_@
+
+            void @_complete_@() { // @_exposition-only_@
+                set_value(std::move(@_receiver_@));
+            }
+        };
+
+        static constexpr auto @_get-state_@ =
+          []<class Receiver>(auto&& sender, Receiver& receiver) {
+            auto[_, self] = sender;
+            return @_state_@<Receiver>(self, receiver);
+          };
+        static constexpr auto @_start_@ =
+            [](auto& s, auto&) { @_see-below_@; };
+    };
+```
+
+[7]{.pnum} In the function object used to initialize
+    `@_impls-for_@<@_join-t_@>::@_start_@` let `state` be `s.@_scope_@->@_state_@`.
+    If `state` is
+
+- [7.1]{.pnum} `@_unused_@`, `@_unused-and-closed_@`, or `@_joined_@`,
+    `s.@_complete_@()` is invoked;
+- [7.2]{.pnum} `@_open_@`, changes the state of `*s.@_scope_@` to `@_open-and-joining_@`;
+- [7.3]{.pnum} `@_closed_@`, changes the state of `*s.@_scope_@` to `@_closed-and-joining_@`;
+
+[8]{.pnum} If `s.@_complete_@()` was not invoked, registers `s` with
+    `*s.@_scope_@` to have `s.@_complete_@()` invoked when `s.@_scope_@->@_count_@` becomes
+    zero.
```

Comment:
I'm not sure I understand how _`impls-for`_ works, but I don't think this is sufficient; the join-sender should complete on the execution context associated with `get_scheduler(receiver)` if it doesn't complete inline in `start()`.

Also, I think that starting the join-sender in the _`unused`_ or _`unused-and-closed`_ state should probably move the state to _`joined`_ so that no future `try_associate()` calls can succeed.

```suggestion
    template <>
    struct @_impls-for_@<@_join-t_@>: @_default-impls_@> {
        template <class Receiver>
        struct @_state_@ {  // @_exposition-only_@
            simple_counting_scope* @_scope_@; // @_exposition-only_@
            remove_cvref_t<Receiver>& @_receiver_@; // @_exposition-only_@
            using @_op_t_@ = decltype(connect(schedule(get_scheduler(receiver)), receiver)); // @_exposition-only_@
            op_t @_op_@; // @_exposition-only_@

            @_state_@(simple_counting_scope* scope, Receiver& receiver) // @_exposition-only_@
              : @_scope_@(scope),
                @_receiver_@(receiver),
                @_op_@(connect(schedule(get_scheduler(receiver)), receiver)) {}

            void @_complete_@() { // @_exposition-only_@
                @_op_@.start();
            }

            void @_complete-inline_@() { // @_exposition-only_@
                set_value(std::move(@_receiver_@));
            }
        };

        static constexpr auto @_get-state_@ =
          []<class Receiver>(auto&& sender, Receiver& receiver) {
            auto[_, self] = sender;
            return @_state_@<Receiver>(self, receiver);
          };
        static constexpr auto @_start_@ =
            [](auto& s, auto&) { @_see-below_@; };
    };
```

[7]{.pnum} In the function object used to initialize
    `@_impls-for_@<@_join-t_@>::@_start_@` let `state` be `s.@_scope_@->@_state_@`.
    If `state` is

- [7.1]{.pnum} `@_unused_@`, `@_unused-and-closed_@`, or `@_joined_@`,
    `s.@_complete-inline_@()` is invoked and changes the state of `*s.@_scope_@` to `@_joined_@`;
- [7.2]{.pnum} `@_open_@`, changes the state of `*s.@_scope_@` to `@_open-and-joining_@`;
- [7.3]{.pnum} `@_closed_@`, changes the state of `*s.@_scope_@` to `@_closed-and-joining_@`;

[8]{.pnum} If `s.@_complete-inline_@()` was not invoked, registers `s` with
    `*s.@_scope_@` to have `s.@_complete_@()` invoked when `s.@_scope_@->@_count_@` becomes
    zero.
```

Reply:
Yes, I didn't have on the radar that the join-sender needs to take the result of `get_scheduler(get_env(receiver))` into account. I think that change makes sense. When I looked at it I first thought that returning `continues_on(some-sender)` would be the thing but I like your approach better.

Note that there is a `get_env(receiver)` needed: the receivers are not required to respond to queries. I _think_ receiver can be their own environments but generic code need to travel via the environment.